### PR TITLE
Add requirements for the spec-file library

### PR DESCRIPTION
**Pull request by lachmanfrantisek**
Signed-off-by: Frantisek Lachman <flachman@redhat.com>

### CODE REVIEW

File: specfile/README.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
+# Packit requirements for the spec-file library
+
+## Manipulation
+
+- We create a Python object by specifying a path of the spec-file.
```
Comment (TomasTomecek) [inline specfile/README.md:5]: I know that we have to specify the path to spec file sources - should we mention that?

File: specfile/README.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+- (optional) After a change, we can reload the content (can be done by recreating of the object).
+- After the manipulation, the content is synced to the file or, we have a way to do this explicitely.
+- Be able to copy a section of spec-file to another object.
+
+## Version and release
+
+- The version and release fields can be get.
+- We can set the new version. If release is not specified, reset the release number to 1.
```
Comment (TomasTomecek) [inline specfile/README.md:13]:
```suggestion
- We can set the new version. If release is not specified, reset the release number to 1 while preserving all the macros.
```

**Review by TomasTomecek — APPROVED** [review]

File: specfile/README.md (hunk 2 above)
Comment (lachmanfrantisek) [reply specfile/README.md:13]: What about mentioning this in the beginning that we want to save only relevant changes, not the expanded macros.

File: specfile/README.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
+# Packit requirements for the spec-file library
+
+## Manipulation
+
+- We create a Python object by specifying a path of the spec-file.
+- (optional) After a change, we can reload the content (can be done by recreating of the object).
+- After the manipulation, the content is synced to the file or, we have a way to do this explicitely.
```
Comment (csomh) [inline specfile/README.md:7]:
```suggestion
- After the manipulation, the content is synced to the file or, we have a way to do this explicitly.
```

File: specfile/README.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+- Be able to copy a section of spec-file to another object.
+
+## Version and release
+
+- The version and release fields can be get.
```
Comment (csomh) [inline specfile/README.md:12]: Do you mean this with macros expanded or the raw values? Or both?

File: specfile/README.md [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+- We can set the new version. If release is not specified, reset the release number to 1.
+
+## Changelog
+
+- We can get the content of the changelog.
+- We can add a new changelog entry:
+  - By providing only the changelog text.
+  - Optionally, field like `name` or `email` can be specified.
```
Comment (csomh) [inline specfile/README.md:20]: What values are going to be used, if no name and email is specified?

File: specfile/README.md [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+- We can get the sources and patches.
+- We can download the sources from a lookaside cache.
+- We can download remote sources (sources in form of URL).
+- We can change the value of a source.
+- We can get archive name = basename of the first Source in SPEC file a.k.a. `Source`/`Source0`.
+
+## Comments
+
+- For patches and other fields, we want to add a comment without a need to manipulate with the file directly.
```
Comment (csomh) [inline specfile/README.md:32]: :+1:

File: specfile/README.md [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+  - By providing only the changelog text.
+  - Optionally, field like `name` or `email` can be specified.
+
+## Sources
+
+- We can get the sources and patches.
+- We can download the sources from a lookaside cache.
+- We can download remote sources (sources in form of URL).
+- We can change the value of a source.
+- We can get archive name = basename of the first Source in SPEC file a.k.a. `Source`/`Source0`.
```
Comment (csomh) [inline specfile/README.md:28]: What about the others? Why just the first one?

**Review by csomh — COMMENTED** [review]
I think it should also be clarified for each use-case (or in general) if we want those operations with macros expanded or on the raw text in the spec-file.

File: specfile/README.md (hunk 7 above)
Comment (lachmanfrantisek) [reply specfile/README.md:28]: I've got that from the docstring of the rebase-helper method..;) I'll remove the `first`...

File: specfile/README.md (hunk 4 above)
Comment (lachmanfrantisek) [reply specfile/README.md:12]: This is with what I am really not sure in general -- do we want to have both modes?

For us, it's crucial to be able to do the edits without side effects (the diff after a change is local and minimal).

File: specfile/README.md (hunk 2 above)
Comment (lachmanfrantisek) [reply specfile/README.md:14]: I've added `All changes results in a minimal and local diff.` to the very beginning.

File: specfile/README.md (hunk 1 above)
Comment (lachmanfrantisek) [reply specfile/README.md]: Is this right (mainly the default)?
```
- We create a Python object by specifying a path of the spec-file and (optionally) path to spec-file sources
  (the directory with the spec-file).
```

**Comment by lachmanfrantisek** [comment]
@TomasTomecek @csomh I've included your suggestion but am still not sure about the macros.

What do you think about that?

Here are my points:
* Not expanding macros can lead to non-parsable spec-file.
* Expanding macros can make it hard to edit the file with a minimal diff.
* I am thinking about supporting both modes but making it read-only when macros are expanded. Or just clearly state that the final diff can be weird.

**Comment by csomh** [comment]
> @TomasTomecek @csomh I've included your suggestion but am still not sure about the macros.

Can we think about this topic as: *we want an API to be able to edit spec-files similar to how package developer do it*?

If I put it this way, than in some cases—mostly when editing the files (?)—the API should provide access to the raw spec-file, as this is what a package developer would also do. 

In other cases—mostly when reading the values of tags (?)—we are interested in the values with macros expanded (reading `Source0` for example which often contains macros).

Following the train of though above I would say that the methods which edit the spec-file should operate on the raw-spec, eventually doing some kind of validation after editing that the spec still can be parsed. Reading methods should have an option to either get things with macros expanded or not. The latter would be probably handy when implementing the methods to edit the spec.

**Comment by lachmanfrantisek** [comment]
@csomh I agree, both modes makes sense only for reading. I've included this in the document. I hope we are done for now and can go deeper together with other teams.

**Review by csomh — APPROVED** [review]

**Pull request merged**